component/service/src/main/java/org/exoplatform/social/rest/impl/activity/ActivityRestResourcesV1.java

```diff
@@ -855,6 +855,65 @@ public Response deleteLike(
     return getLikesOfActivity(uriInfo, activityId, 0, RestUtils.DEFAULT_LIMIT, null);
   }
 
+  @POST
+  @Path("{activityId}/pins")
+  @Produces(MediaType.APPLICATION_JSON)
+  @RolesAllowed("users")
+  @Operation(summary = "Pin a specific activity to space stream", method = "POST", description = "This pins an activity to space stream if the authenticated user is a manager or a redactor of the space.")
+  @ApiResponses(value = { @ApiResponse(responseCode = "200", description = "Request fulfilled"),
+      @ApiResponse(responseCode = "500", description = "Internal server error"),
+      @ApiResponse(responseCode = "400", description = "Invalid query input") })
+  public Response pinActivity(@Context UriInfo uriInfo,
+                              @Context Request request,
+                              @Parameter(description = "Activity id", required = true) @PathParam("activityId") String activityId) {
+
+    if (StringUtils.isBlank(activityId)) {
+      return Response.status(Response.Status.BAD_REQUEST).build();
+    }
+    org.exoplatform.services.security.Identity authenticatedUserIdentity = ConversationState.getCurrent().getIdentity();
+    String authenticatedUser = authenticatedUserIdentity.getUserId();
+    Identity currentUser = identityManager.getOrCreateIdentity(OrganizationIdentityProvider.NAME, authenticatedUser);
+
+    ExoSocialActivity activity = activityManager.getActivity(activityId);
+
+    if (!activityManager.canPinActivity(activity, currentUser)) {
```

Comment:
What if activity is null (non existing id) ?

Reply:
It seems to generate a NPE.

Reply:
Example on builders : 

> 2022-10-18 08:28:24,913 | WARN  | Uncaught REST Service Exception [o.e.s.rest.impl.DefaultExceptionMapper<http-nio-0.0.0.0-8080-exec-144>] 
> java.lang.NullPointerException: null
> 	at org.exoplatform.social.core.manager.ActivityManagerImpl.canPinActivity(ActivityManagerImpl.java:569)
> 	at org.exoplatform.social.rest.impl.activity.ActivityRestResourcesV1.pinActivity(ActivityRestResourcesV1.java:879)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> 	at org.exoplatform.services.rest.impl.method.DefaultMethodInvoker$1.run(DefaultMethodInvoker.java:192)
> 	at org.exoplatform.commons.utils.SecurityHelper.doPrivilegedExceptionAction(SecurityHelper.java:311)
> 	at org.exoplatform.services.rest.impl.method.DefaultMethodInvoker.invokeMethod(DefaultMethodInvoker.java:188)
> 	at org.exoplatform.services.rest.impl.method.DefaultMethodInvoker.invokeMethod(DefaultMethodInvoker.java:181)
> 	at org.exoplatform.services.rest.impl.RequestDispatcher.invokeSubResourceMethod(RequestDispatcher.java:657)
> 	at org.exoplatform.services.rest.impl.RequestDispatcher.dispatch(RequestDispatcher.java:602)
> 	at org.exoplatform.services.rest.impl.RequestDispatcher.dispatch(RequestDispatcher.java:507)
> 	at org.exoplatform.services.rest.impl.RequestHandlerImpl.handleRequest(RequestHandlerImpl.java:143)
> 	at org.exoplatform.services.rest.servlet.RestServlet.onService(RestServlet.java:100)
> 	at org.exoplatform.container.web.AbstractHttpServlet.service(AbstractHttpServlet.java:133)
> 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:764)

I used pins endpoint, with activityId = 100000 : 
`romain@romain-XPS-ubuntu:~$ curl -X POST -b cookies https://builders.meeds.io/portal/rest/v1/social/activities/100000/pins
java.lang.NullPointerException`

Reply:
I have a security question here : 
When I call this endpoint with an activity id : 
- if it answer 404, i know that the activity id do not exists
- if it answer 403, i know that activity id exists and that I'm not authorized to pin it.
So we disclose information about activity id.

In function getActivityById, we answer 404 in case activity not exists, or in case activity is not viewable by me.

We have two options : 
- if activity id not exists and if i'm not authorized to pin it, answer 404
- if activity id not exists, and if i'm not authorized to view it, answer 404, if i'm authorized to see it, but not to pin it, answer 403.
Second case is perhaps more complex to do.

Same remark for unpin function.

WDYT @boubaker ?

Reply:
> if activity id not exists, and if i'm not authorized to view it, answer 404, if i'm authorized to see it, but not to pin it, answer 403.

I agree

Reply:
> Second case is perhaps more complex to do.

Not too much in fact, it will be just same condition as getActivityById

```java
if (activity == null || !activityManager.isActivityViewable(activity, authenticatedUserIdentity)) {
  throw new WebApplicationException(Response.Status.NOT_FOUND);
}

```

Reply:
:+1:

Reply:
Done!